### DISCUSSION TITLE
How to make platform-specific implementations for Linux, which seems to have no predefined platform symbol?

### DISCUSSION THREAD

**Discussion by hansmbakker** [unoplatform/uno#9446] [Q&A]
I'm trying to create a cross-platform library to be consumed in an Uno app. I used the Uno library project template.

As explained on https://platform.uno/docs/articles/platform-specific-csharp.html, **some** platforms have predefined symbols that can be used to create platform-specific implementations of a class using a statement like `#if __ANDROID__`.

However, there seems to be a gap between the predefined symbols and the platforms that Uno supports.

How can I implement the platform-specific part for Linux?
And - how to use the Uno library project template? Will there need to be a separate build per platform or is the resulting DLL reusable across platforms?

**Comment by jeromelaban** [comment]
There is no "linux" specific build as far as Uno is concerned.

It is possible to make a ["cross-targeted library"](https://platform.uno/docs/articles/cross-targeted-libraries.html?q=cross), which allows to make skia-backend specific code, but that code will be valid for any platform that can run the skia backend (Windows, macOS, Linux, Tizen). We generally use this type of library only when there's a need for very specific nuget dependencies that may incorrectly flow through dependents.

If you want code to be executed on linux specifically, you'll need to use [`IsOSPlatform`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.runtimeinformation.isosplatform?view=net-6.0#system-runtime-interopservices-runtimeinformation-isosplatform(system-runtime-interopservices-osplatform)) and enable code at runtime.

**Reply by hansmbakker** [reply]
Thank you for your reply! The usecase was to make the code in https://github.com/dotnet/maui/tree/main/src/Essentials/src/SecureStorage available for an Uno app.

Given the platform-specific dependencies and implementation differences, it is needed to have platform-specific code.

Maui does this using conditional file includes, but that only works because they do not support Linux as a separate platform I guess.

I also thought about a workaround like this, but likely it is brittle?

```csharp
#if __ANDROID__ || __IOS__ || __MACOS__ || NETFX_CORE
#elif
   // put code here
# endif
```
And, WinUI likely does not use NETFX_CORE?

**Reply by jeromelaban** [reply]
Yes it's possible to do platform specific code, in various ways: https://platform.uno/docs/articles/platform-specific-csharp.html. And indeed, MAUI does not have the Linux problem, because all target frameworks map to individual platforms, which is not case for Uno as more targets platforms map to individual targets frameworks.
